File: data/code_corpus/c/linked_list.c
/*
 * linked_list.c
 * Teaches: a singly-linked list of ints with head/tail insert and full
 * teardown. Shows malloc + free pairing, struct self-reference, and
 * the head-pointer pattern. Intermediate sample.
 * Build: gcc -Wall -o linked_list linked_list.c
 * Run:   ./linked_list
 */

#include <stdio.h>
#include <stdlib.h>

/* `struct node` is self-referential: each node holds an int and a
 * pointer to the next node in the chain (or NULL at the end). */
struct node {
    int value;
    struct node *next;
};

/* Allocate a single node on the heap. Caller is responsible for
 * eventually free()-ing it (typically via list_free below). */
static struct node *node_new(int value)
{
    struct node *n = (struct node *)malloc(sizeof(*n));
    if (n == NULL) {
        /* Out of memory. A teaching-grade program just exits; a real
         * library would propagate an error code. */
        fprintf(stderr, "out of memory\n");
        exit(1);
    }
    n->value = value;
    n->next = NULL;
    return n;
}

/* Prepend `value`. The list is identified by a pointer-to-pointer
 * because we may need to update the head itself. */
static void list_insert_head(struct node **head, int value)
{
    struct node *n = node_new(value);
    n->next = *head;   /* old head becomes second node */
    *head = n;
}

/* Append `value`. O(n) because we walk to the last node. */
static void list_insert_tail(struct node **head, int value)
{
    struct node *n = node_new(value);
    if (*head == NULL) {
        *head = n;
        return;
    }
    struct node *cursor = *head;
    while (cursor->next != NULL) {
        cursor = cursor->next;
    }
    cursor->next = n;
}

/* Print the values "[a, b, c]"-style. */
static void list_print(const struct node *head)
{
    printf("[");
    const struct node *cursor = head;
    while (cursor != NULL) {
        printf("%d", cursor->value);
        if (cursor->next != NULL) {
            printf(", ");
        }
        cursor = cursor->next;
    }
    printf("]\n");
}

/* Free every node in the chain. After this call, the caller's head
 * pointer is dangling; we set it to NULL via the pointer-to-pointer. */
static void list_free(struct node **head)
{
    struct node *cursor = *head;
    while (cursor != NULL) {
        struct node *doomed = cursor;
        cursor = cursor->next;
        free(doomed);  /* always pair every malloc with exactly one free */
    }
    *head = NULL;
}

int main(void)
{
    struct node *head = NULL;

    /* Build [1, 2, 3] by appending. */
    list_insert_tail(&head, 1);
    list_insert_tail(&head, 2);
    list_insert_tail(&head, 3);
    printf("after appends: ");
    list_print(head);

    /* Prepend 0 -> [0, 1, 2, 3]. */
    list_insert_head(&head, 0);
    printf("after head insert: ");
    list_print(head);

    /* Free every node so we don't leak. */
    list_free(&head);
    printf("after free: ");
    list_print(head);   /* [] */

    return 0;
}

File: data/code_corpus/python/linked_list.py
# linked_list.py
# Teaches: a singly-linked list as a class, with insert/traverse/delete.
# Demonstrates head pointer, node chaining, and length tracking.
# Intermediate sample. Run: python3 linked_list.py


class Node:
    """One link in the chain: a value and a reference to the next node."""

    def __init__(self, value):
        self.value = value
        # `None` is Python's null. The last node's `next` is None,
        # which is how we recognize the end of the list.
        self.next = None


class LinkedList:
    """A singly-linked list with O(1) head insert and O(n) tail insert.

    We track `length` explicitly so `len()` is O(1); without it we'd
    have to walk the chain every time.
    """

    def __init__(self):
        self.head = None
        self.length = 0

    def __len__(self):
        # Implementing __len__ lets callers use the built-in len().
        return self.length

    def insert_head(self, value):
        """Prepend `value`. Constant-time: only the head changes."""
        new_node = Node(value)
        new_node.next = self.head  # old head becomes the second node
        self.head = new_node
        self.length += 1

    def insert_tail(self, value):
        """Append `value`. Linear-time: must walk to the last node."""
        new_node = Node(value)

        # Empty-list special case: the new node IS the head.
        if self.head is None:
            self.head = new_node
            self.length += 1
            return

        # Walk until we find the node whose `next` is None.
        cursor = self.head
        while cursor.next is not None:
            cursor = cursor.next
        cursor.next = new_node
        self.length += 1

    def delete_first(self, value):
        """Remove the first node holding `value`. Returns True on success."""
        # Empty list: nothing to delete.
        if self.head is None:
            return False

        # Special-case the head: there's no predecessor to rewire.
        if self.head.value == value:
            self.head = self.head.next
            self.length -= 1
            return True

        # General case: walk with a "previous" pointer so we can splice
        # the target out by setting prev.next = target.next.
        prev = self.head
        cursor = self.head.next
        while cursor is not None:
            if cursor.value == value:
                prev.next = cursor.next
                self.length -= 1
                return True
            prev = cursor
            cursor = cursor.next
        return False

    def to_list(self):
        """Return the values as a regular Python list (helpful for tests)."""
        out = []
        cursor = self.head
        while cursor is not None:
            out.append(cursor.value)
            cursor = cursor.next
        return out


def main():
    ll = LinkedList()

    # Build [10, 20, 30] by appending.
    for v in (10, 20, 30):
        ll.insert_tail(v)
    print("after appends:", ll.to_list(), "len=", len(ll))

    # Prepend 5 -> [5, 10, 20, 30].
    ll.insert_head(5)
    print("after head insert:", ll.to_list())

    # Delete a middle value.
    ok = ll.delete_first(20)
    print("delete 20 ok=", ok, "->", ll.to_list())

    # Delete the head.
    ok = ll.delete_first(5)
    print("delete 5  ok=", ok, "->", ll.to_list())

    # Try to delete something not present.
    ok = ll.delete_first(999)
    print("delete 999 ok=", ok, "->", ll.to_list())


if __name__ == "__main__":
    main()
